### PR TITLE
Allow WeatherOverlay fade in/out when enabled/disabled

### PR DESCRIPTION
**Pull request by dnqbob**

### CODE REVIEW

File: OpenRA.Mods.Common/Scripting/Global/WorldActorConditionGlobal.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Linq;
+using Eluant;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Scripting;
+
+namespace OpenRA.Mods.Common.Scripting
+{
+	[ScriptGlobal("WorldActor")]
+	public class WorldActorConditionGlobal : ScriptGlobal
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Global/WorldActorConditionGlobal.cs]: This is mostly redundant against `ConditionProperties`. You just need to add a `WorldActor` property to `MapGlobal`.

Comment (dnqbob) [reply OpenRA.Mods.Common/Scripting/Global/WorldActorConditionGlobal.cs]: `WorldActorConditionGlobal` is just part of testcase, used for making the test of some edge cases easily (for example, fade out when fading in is not finished), it is unrelated to the actual changing. 

We can add that property in another PR, though.

File: OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs [hunk 2]
```diff
@@ -159,30 +174,60 @@ void INotifyViewportZoomExtentsChanged.ViewportZoomExtentsChanged(float minZoom,
...
+				fadingOutTick = (fadeInTick - fadingInTick) * 1000 * fadeOutTick / fadeInTick / 1000;
 		}
 
 		void ITick.Tick(Actor self)
 		{
-			if (!info.ChangingWindLevel || info.WindLevels.Length == 1)
+			if (initialTick)
+			{
+				initialTick = false;
+				var initialPercentage = Info.InitialParticlePercentage > 100 ? 100 : (Info.InitialParticlePercentage < 0 ? 0 : Info.InitialParticlePercentage);
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs]: `var initialPercentage = Exts.Clamp(Info.InitialParticlePercentage, 0, 100);`

**Review by RoosterDragon — APPROVED** [review]
One fixup, otherwise LGTM.

File: OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs (hunk 2 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs]: Done fixup.

File: OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs [hunk 3]
```diff
@@ -69,10 +69,19 @@ public class WeatherOverlayInfo : TraitInfo, ILobbyCustomRulesIgnore
...
 		public readonly byte LineTailAlphaValue = 200;
 
+		[Desc("How long the time the weather fades in when enabled.")]
+		public readonly int FadeOutTicks = 1000;
+
+		[Desc("How long the time the weather fades out when disabled.")]
+		public readonly int FadeInTicks = 1000;
+
+		[Desc("Percentage of the initial particle when enabled and the game start.")]
+		public readonly int InitialParticlePercentage = 100;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:79]: This feels a bit speculatively general IMO. Do we have a realistic usecase for needing a value that isn't 100 when it starts enabled and 0 when it starts disabled?

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:79]: This option is used for when player want to start from mid-stage of the weather, any weather. For example, a map starts from mid-stage of a thundershower: there is only mid rain but get heavier as time goes by, when it at max stage there will be lightning and thunder by using `FlashPostProcessEffect` (we can make it conditional and scriptable later).

File: OpenRA.Mods.Common/Scripting/Global/WorldActorConditionGlobal.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Scripting/Global/WorldActorConditionGlobal.cs]: I filled #21188 on `ConditionProperties` on WorldActor

File: OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs (hunk 3 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:79]: I think it makes sense to have this option

File: OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs [hunk 4]
```diff
@@ -69,10 +69,19 @@ public class WeatherOverlayInfo : TraitInfo, ILobbyCustomRulesIgnore
 		[Desc("Works only with line enabled and can be used to fade out the tail of the line like a contrail.")]
 		public readonly byte LineTailAlphaValue = 200;
 
+		[Desc("How long the time the weather fades in when enabled.")]
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs]:
```suggestion
		[Desc("Time to fade in once the trait becomes enabled.")]
```

to be honest these names feel reverse as well

File: OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs [hunk 5]
```diff
@@ -69,10 +69,19 @@ public class WeatherOverlayInfo : TraitInfo, ILobbyCustomRulesIgnore
...
+		public readonly int FadeOutTicks = 1000;
+
+		[Desc("How long the time the weather fades out when disabled.")]
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs]:
```suggestion
		[Desc("Time to fade out once the trait becomes disabled.")]
```

File: OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs (hunk 4 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs]: fixed

File: OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs (hunk 5 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs]: fixed

**Review by PunkPun — APPROVED** [review]
LGTM

**Comment by PunkPun** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/58fa46dd4d6b675f77dc0a231a20ea1bb802e102...6f3c3e473557b9a36116a8e683e504f557eead84)
